src/realm/util/encrypted_file_mapping.cpp

```diff
@@ -669,10 +684,34 @@ void EncryptedFileMapping::refresh_page(size_t local_page_ndx, size_t required)
     char* addr = page_addr(local_page_ndx);
 
     if (!copy_up_to_date_page(local_page_ndx)) {
+
         size_t page_ndx_in_file = local_page_ndx + m_first_page;
+        off_t data_pos = off_t(page_ndx_in_file << m_page_shift);
+        if (is(m_page_state[local_page_ndx], StaleIV)) {
+            auto refreshed_ivs = m_file.cryptor.refresh_ivs(m_file.fd, data_pos, page_ndx_in_file);
+            for (const auto& [page_ndx, state] : refreshed_ivs) {
+                size_t local_page_ndx_of_iv_change = page_ndx - m_first_page;
+                if (!contains_page(page_ndx) || is_not(m_page_state[local_page_ndx_of_iv_change], StaleIV)) {
+                    continue;
+                }
+                clear(m_page_state[local_page_ndx_of_iv_change], StaleIV);
+                switch (state) {
+                    case IVRefreshState::UpToDate:
+                        set(m_page_state[local_page_ndx_of_iv_change], UpToDate);
+                        break;
+                    case IVRefreshState::RequiresRefresh:
+                        clear(m_page_state[local_page_ndx_of_iv_change], UpToDate);
```

Comment:
I don't think we can ever have both StaleIV and UpToDate set to begin with? We clear UpToDate when setting StaleIV.